WordPress/gutenberg
# Block settings extensibility

Add a Slot in the BlockSettingsMenu for third-parties to hook into.

* `onClick` will be called when the user clicks the menu item.
* `allowedBlocks` should be a whitelist of block names. If not present the item will be shown for every block. If multiple blocks are selected, the item will be shown if all of them are in the whitelist.

An example on how a plugin would use this API:

```jsx
import { registerPlugin } from '@wordpress/plugins';
import { PluginBlockSettingsMenuItem } from '@wordpress/editPost';

const doOnClick = ( ) => {
    // the plugin doing its dance
};

registerPlugin( 'plugin-name', {
    render: ( ) => (
        <PluginBlockSettingsMenuItem 
            allowedBlockNames=[ 'core/paragraph' ]
            icon='dashicon-name'
            label='Plugin name'
            onClick={ doOnClick } />
    ),
} );
```

### ~Open questions~

- [x] ~should the API allow whitelist/blacklist the block types in which this item should be shown? If so, what would be the best way to declare this? I was thinking about adding a new prop `blockList` to the `BlockSettingsMenuPluginsItem` that would contain the block names.~ Mimic the `allowedBlocks` property of [InnerBlocks](https://github.com/WordPress/gutenberg/blob/803c12a9f4b2ffb33151c2a0c0305b1ac731989f/editor/components/inner-blocks/README.md).
- [x] ~is the block UID enough info for the callback or do we anticipate the plugin would need anything else?~ Plugin authors should use the selectors in core data package to get access to the selected blocks.

### Testing

* Download and activate the [test plugin](https://github.com/nosolosw/blocksettingsmenu-plugin).
* Make sure a new item in the settings menu is shown for paragraph blocks, and the icon is `screenoptions`. On clicking, `Block clicked` is logged to the console.
* Modify the `allowedBlocks` property (no prop, different block names, multiple selections, etc) and test that results are expected.
